src/vm/methodtable.cpp

```diff
@@ -7059,24 +7117,62 @@ BOOL MethodTable::FindDefaultMethod(
                         );
                     }
 
-                    if (pBestCandidateMT != pCurMT)
-                    {                    
-                        if (pBestCandidateMT == NULL ||                         // first time
-                            pCurMT->CanCastToInterface(pBestCandidateMT))       // Prefer "more specific"" interface
+                    bool needToInsert = true;
+                    bool seenMoreSpecific = false;
+
+                    // We need to maintain the invariant that the candidates are always the most specific
+                    // in all path scaned so far. There might be multiple incompatible candidates 
+                    for (int i = 0; i < candidatesCount; ++i)
+                    {
+                        MethodTable *pCandidateMT = candidates[i].pMT;
+                        if (pCandidateMT == NULL)
+                            continue;
+
+                        if (pCandidateMT->HasSameTypeDefAs(pCurMT))
```

Comment:
HasSameTypeDefAs will return true even when 2 instantiations are different, if the generic typedef is the same.

---
In reply to: [132596188](https://github.com/dotnet/coreclr/pull/12753#discussion_r132596188) [](ancestors = 132596188)